scripts/Inelastic/IndirectCommon.py

```diff
@@ -188,31 +188,38 @@ def createQaxis(inputWS):
 
 
 def GetWSangles(inWS):
-    num_hist = s_api.mtd[inWS].getNumberHistograms()  # get no. of histograms/groups
-    source_pos = s_api.mtd[inWS].getInstrument().getSource().getPos()
-    sample_pos = s_api.mtd[inWS].getInstrument().getSample().getPos()
+    if isinstance(inWS, str):
+        ws = s_api.mtd[inWS]
+    else:
+        ws = inWS
+    num_hist = ws.getNumberHistograms()  # get no. of histograms/groups
+    source_pos = ws.getInstrument().getSource().getPos()
+    sample_pos = ws.getInstrument().getSample().getPos()
     beam_pos = sample_pos - source_pos
     angles = []  # will be list of angles
     for index in range(0, num_hist):
-        detector = s_api.mtd[inWS].getDetector(index)  # get index
+        detector = ws.getDetector(index)  # get index
         two_theta = detector.getTwoTheta(sample_pos, beam_pos) * 180.0 / math.pi  # calc angle
         angles.append(two_theta)  # add angle
     return angles
 
 
-def GetThetaQ(ws):
+def GetThetaQ(ws_in):
     """
     Returns the theta and elastic Q for each spectrum in a given workspace.
 
     @param ws Workspace to get theta and Q for
     @returns A tuple containing a list of theta values and a list of Q values
     """
-
+    if isinstance(ws_in, str):
```

Comment:
Something similar is handled nicely in ASE using decorators. There are lots of reader and writer functions that take a target file as an argument. This could be a string, a Path object or an open file handle: `@reader` and `@writer` decorators are provided that process this so the inside of the function always sees a file handle.

https://gitlab.com/ase/ase/-/blob/master/ase/utils/__init__.py?ref_type=heads#L469